boards: arm: Add support for ST B-L072Z-LRWAN1 board

deviance opened the pull request:
Added support for ST B-L072Z-LRWAN1 board configuration.
This board uses the same MCU as Dragino-LSN50.

Signed-off-by: Aleksandr Makarov <aleksandr.o.makarov@gmail.com>

deviance committed:
boards: arm: Add support for ST B-L072Z-LRWAN1 board

Added support for ST B-L072Z-LRWAN1 board configuration.
This board uses the same MCU as Dragino-LSN50.

Signed-off-by: Aleksandr Makarov <seems.deviant@gmail.com>

erwango reviewed (changes requested):
Following #6858 and #7151, we're trying to enforce
boards to provide minimum board configuration.
Since it appears that your board provides an arduino connector, please populate and enable matching connectors. 
Cf #9915 as example.

Aim is to increase feature set on available boards, extent CI testing, and provide upfront compatibility with recently introduced shield feature (#6837)

dbkinder commented on boards/arm/b_l072z_lrwan1/doc/b_l072z_lrwan1.rst:
```diff
@@ -0,0 +1,225 @@
+
+- CMWX1ZZABZ-091 LoRa |reg| /Sigfox |trade| module (Murata)
+
+        - Embedded ultra-low-power STM32L072CZ Series MCUs, based on Arm |reg| Cortex |reg| -M0+ core, with 192 Kbytes of Flash memory, 20 Kbytes of RAM, 20 Kbytes of EEPROM
```
Try to keep lines <80 chars long:

    - Embedded ultra-low-power STM32L072CZ Series MCUs,
      based on Arm* Cortex* -M0+ core, with 192KB
      of Flash memory, 20KB of RAM, 20KB of EEPROM

dbkinder commented on boards/arm/b_l072z_lrwan1/doc/b_l072z_lrwan1.rst:
```diff
@@ -0,0 +1,225 @@
+
+This kit provides:
+
+- CMWX1ZZABZ-091 LoRa |reg| /Sigfox |trade| module (Murata)
```
We generally recognize third-party trademarks with a simple asterisk instead of a TM or (R), so you can change these references to simply ``LoRa*``, etc.

dbkinder commented on boards/arm/b_l072z_lrwan1/doc/b_l072z_lrwan1.rst:
```diff
@@ -0,0 +1,225 @@
+
+Other hardware features are not yet supported on this Zephyr port.
+
+The default configuration can be found in the defconfig file:
```
This renders better as:

    The default configuration can be found in the defconfig file:
    ``boards/arm/b_l072z_lrwan1/b_l072z_lrwan1_defconfig``

dbkinder commented on boards/arm/b_l072z_lrwan1/doc/b_l072z_lrwan1.rst:
```diff
@@ -0,0 +1,225 @@
+Flashing
+========
+
+B-L072Z-LRWAN1 Discovery board includes an ST-LINK/V2-1 embedded debug tool interface. This interface is supported by the openocd version included in the Zephyr SDK since v0.9.2.
```
keep <80 chars

dbkinder reviewed (changes requested):
Just a few recommended doc changes, thanks!

erwango commented:
Please merge your last fix commit into matching 'guilty' commit.

dbkinder commented on boards/arm/b_l072z_lrwan1/doc/b_l072z_lrwan1.rst:
```diff
@@ -0,0 +1,233 @@
+
+More information about STM32L072CZ can be found here:
+
+        - `STM32L072CZ on www.st.com`_
```
This renders better when the bullet points are not indented:

    More information about STM32L072CZ can be found at:

    - `STM32L072CZ on www.st.com`_
    - `STM32L0x2 reference manual`_

dbkinder reviewed (changes requested):
Looks good, one small edit

deviance commented:
Broken commit history by previous push, but now everything is fine.

erwango commented:
@deviance , can get CI pass?

deviance commented:
> @deviance , can get CI pass?

No, I couldn't understand where "#include <st/stm32l072.dtsi>" error comes from.

erwango commented:
@deviance 

> No, I couldn't understand where "#include <st/stm32l072.dtsi>" error comes from.

There has been some changes in dts/st.
You should use:
```
#include <st/l0/stm32l072Xz.dtsi>
```

deviance commented:
> @deviance
> 
> > No, I couldn't understand where "#include <st/stm32l072.dtsi>" error comes from.
> 
> There has been some changes in dts/st.
> You should use:
> 
> ```
> #include <st/l0/stm32l072Xz.dtsi>
> ```

Rebased to upstream master and updated the .dts include. Tests should pass now.

deviance commented:
> > @deviance
> > > No, I couldn't understand where "#include <st/stm32l072.dtsi>" error comes from.
> > 
> > 
> > There has been some changes in dts/st.
> > You should use:
> > ```
> > #include <st/l0/stm32l072Xz.dtsi>
> > ```
> 
> Rebased to upstream master and updated the .dts include. Tests should pass now.

@erwango I'm ignoring 'samples' tag for tests and so tests pass.

erwango commented on boards/arm/b_l072z_lrwan1/b_072z_lrwan1.dts:
```diff
@@ -0,0 +1,32 @@
+		zephyr,sram = &sram0;
+		zephyr,flash = &flash0;
+	};
+};
```
Please add support for user button (B1) and leds(LD1>LD4).

deviance replied:
Added support for those

erwango reviewed (changes requested):
Seems good, but one methodological comment:
We won't merge commits that are known to contain errors even if fix is available in the very next commit. Hence you should not push commits + fixes, but initially clean commits.
So please 'squash' your 2 commits into one in order to obtain clean commit.
Then, if you feel so,  you can split the PR into several logical commits (initial board support, add i2c, add spi, ...), which is easier for review, cf following example:
https://github.com/zephyrproject-rtos/zephyr/pull/10419/commits

erwango commented on boards/arm/b_l072z_lrwan1/b_l072z_lrwan1_defconfig:
```diff
@@ -0,0 +1,43 @@
+CONFIG_UART_CONSOLE=y
+
+# enable SPI
+CONFIG_SPI=y
```
Remove this line. 
Default board configuration should be disabled by default (except basics and always requested: clock/gpio/console).
Others should be enabled by application

erwango commented on boards/arm/b_l072z_lrwan1/b_l072z_lrwan1_defconfig:
```diff
@@ -0,0 +1,43 @@
+CONFIG_GPIO=y
+
+# enable I2C
+CONFIG_I2C=y
```
Remove this line

erwango commented on boards/arm/b_l072z_lrwan1/b_l072z_lrwan1.dts:
```diff
@@ -0,0 +1,87 @@
+	clock-frequency = <I2C_BITRATE_FAST>;
+};
+
+arduino_usart: &usart1 {};
```
Nitpick: Move these lines just before &usart1..
Just to do like others..

erwango reviewed (changes requested):
Last changes and we're good

deviance commented:
@erwango There is an SX1276 LoRa module present on the board and It is available through the SPI1.
Is there any method in a .dts to describe that module as a default networking device? So that the interface become accessible by 'net'-tagged test.

erwango commented:
@deviance 

> @erwango There is an SX1276 LoRa module present on the board and It is available through the SPI1.
Is there any method in a .dts to describe that module as a default networking device? So that the interface become accessible by 'net'-tagged test.

Do we have a driver for SX1276 module?

Anyway, there are two things you need to consider to do so:
-Kconfig.defconfig: Enable matching config flags when NETWORKING is selected, cf how `USB_DEVICE_NETWORK_ECM` is enabled on nucleo_f412zg:
https://github.com/zephyrproject-rtos/zephyr/blob/51574c177f6f2bb671bce4791215f71ad8b83b11/boards/arm/nucleo_f412zg/Kconfig.defconfig#L41
-DTS:  Optionally, if SX1276 driver uses dts information, decalre it as a spi-device on spi1 node. Cf how "st,spbtle-rf" module is declared on disco_l475_iot1 board. This require to provide matching dts yaml binding (`dts/bindings`):
https://github.com/zephyrproject-rtos/zephyr/blob/51574c177f6f2bb671bce4791215f71ad8b83b11/boards/arm/disco_l475_iot1/disco_l475_iot1.dts#L111

deviance commented:
> Do we have a driver for SX1276 module?

@erwango Can't say for sure. The Dragino-LSN50 has this module onboard, but does not define any interfaces to it. Thus it ignores 'net'-tagged tests as well.

deviance commented:
> Anyway, there are two things you need to consider to do so:
> -Kconfig.defconfig: Enable matching config flags when NETWORKING is selected, cf how `USB_DEVICE_NETWORK_ECM` is enabled on nucleo_f412zg:
> [zephyr/boards/arm/nucleo_f412zg/Kconfig.defconfig](https://github.com/zephyrproject-rtos/zephyr/blob/51574c177f6f2bb671bce4791215f71ad8b83b11/boards/arm/nucleo_f412zg/Kconfig.defconfig#L41)
> 
> Line 41 in [51574c1](/zephyrproject-rtos/zephyr/commit/51574c177f6f2bb671bce4791215f71ad8b83b11)
> 
>  if NETWORKING 
> 
> -DTS: Optionally, if SX1276 driver uses dts information, decalre it as a spi-device on spi1 node. Cf how "st,spbtle-rf" module is declared on disco_l475_iot1 board. This require to provide matching dts yaml binding (`dts/bindings`):
> [zephyr/boards/arm/disco_l475_iot1/disco_l475_iot1.dts](https://github.com/zephyrproject-rtos/zephyr/blob/51574c177f6f2bb671bce4791215f71ad8b83b11/boards/arm/disco_l475_iot1/disco_l475_iot1.dts#L111)
> 
> Line 111 in [51574c1](/zephyrproject-rtos/zephyr/commit/51574c177f6f2bb671bce4791215f71ad8b83b11)
> 
>  bt0: spbtle-rf@0 {

I just found out that B-L072Z-LRWAN1 board has a USB interface to LoRa module. USB DP/DM to it are available on PA11/PA12. 

Can this interface be used by default as Ethernet-over-USB?

erwango commented:
@deviance 

> Can this interface be used by default as Ethernet-over-USB?

Do you mean LORA module can be driver as an ethernet device?

erwango commented:
@deviance 

> Lora device to be driven by an Ethernet driver, yes.

From what I understand, B-L072Z-LRWAN1 could be used as a LORA module with an Ethernet-over-USB. Though, to do this you need a driver to route Ethernet packets to SX1276 module over SPI.
So we can configure pins to do this. But it requires to have the ethernet to SPI bridge available, which is not  the case today.
Besides, implementation depends on what type of packets are routed over ethernet (raw IP, AT encapsulated, ..?).
Do you have a pointer to a documentation?

deviance commented:
> @deviance
> 
> > Lora device to be driven by an Ethernet driver, yes.
> 
> From what I understand, B-L072Z-LRWAN1 could be used as a LORA module with an Ethernet-over-USB. Though, to do this you need a driver to route Ethernet packets to SX1276 module over SPI.
> So we can configure pins to do this. But it requires to have the ethernet to SPI bridge available, which is not the case today.
> Besides, implementation depends on what type of packets are routed over ethernet (raw IP, AT encapsulated, ..?).
> Do you have a pointer to a documentation?

The encapsulated protocol is [DASH7](http://www.dash7-alliance.org/product/d7ap1-1/)

erwango commented:
@deviance 

> I just found out that B-L072Z-LRWAN1 board has a USB interface to LoRa module. USB DP/DM to it are available on PA11/PA12.

After further search, I understood that USB interface to LORA module is indeed available.
But this is through PA11/PA12 of the STM32 **inside** module, not the STM32 running Zephyr. So you can't and don't need to configure it. To make it available, you only need to fit matching jumpers (or solder bridge, tbc), which are not fitted by default. No development is required neither.

So for board configuration, you can configure USB, but this is not the same USB that will be connected to LORA module.

erwango commented:
@deviance 

Since requested driver is not available yet, please configure whatever is available today and we'll do more once available

dbkinder reviewed (approved):
+1 for docs

deviance commented:
> @deviance
> 
> Since requested driver is not available yet, please configure whatever is available today and we'll do more once available

As far as I see, no more pins/peripherals can be configured for this board.

I was expecting that PWM2_CH1 could have been connected to PA0 pin, but it seems that `pwm` driver does not have support for STM32L0 timers. Here is what happened when I tried to compile the `blink_led ` sample:
```/home/amakarov/build/zephyr-deviance/drivers/pwm/pwm_stm32.c: In function ‘pwm_stm32_pin_set’:
/home/amakarov/build/zephyr-deviance/drivers/pwm/pwm_stm32.c:97:16: warning: implicit declaration of function ‘IS_TIM_32B_COUNTER_INSTANCE’ [-Wimplicit-function-declaration]
  counter_32b = IS_TIM_32B_COUNTER_INSTANCE(PWM_STRUCT(dev));
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/amakarov/build/zephyr-deviance/drivers/pwm/pwm_stm32.c:118:19: error: ‘TIM_Base_InitTypeDef {aka struct <anonymous>}’ has no member named ‘RepetitionCounter’
  TimerHandle->Init.RepetitionCounter = 0;
                   ^
/home/amakarov/build/zephyr-deviance/drivers/pwm/pwm_stm32.c:129:9: error: ‘TIM_OC_InitTypeDef {aka struct <anonymous>}’ has no member named ‘OCNPolarity’; did you mean ‘OCPolarity’?
  sConfig.OCNPolarity  = TIM_OCNPOLARITY_HIGH;
         ^
/home/amakarov/build/zephyr-deviance/drivers/pwm/pwm_stm32.c:129:25: error: ‘TIM_OCNPOLARITY_HIGH’ undeclared (first use in this function)
  sConfig.OCNPolarity  = TIM_OCNPOLARITY_HIGH;
                         ^~~~~~~~~~~~~~~~~~~~
/home/amakarov/build/zephyr-deviance/drivers/pwm/pwm_stm32.c:129:25: note: each undeclared identifier is reported only once for each function it appears in
/home/amakarov/build/zephyr-deviance/drivers/pwm/pwm_stm32.c:130:9: error: ‘TIM_OC_InitTypeDef {aka struct <anonymous>}’ has no member named ‘OCNIdleState’
  sConfig.OCNIdleState = TIM_OCNIDLESTATE_RESET;
         ^
/home/amakarov/build/zephyr-deviance/drivers/pwm/pwm_stm32.c:130:25: error: ‘TIM_OCNIDLESTATE_RESET’ undeclared (first use in this function)
  sConfig.OCNIdleState = TIM_OCNIDLESTATE_RESET;
                         ^~~~~~~~~~~~~~~~~~~~~~
/home/amakarov/build/zephyr-deviance/drivers/pwm/pwm_stm32.c:131:9: error: ‘TIM_OC_InitTypeDef {aka struct <anonymous>}’ has no member named ‘OCIdleState’
  sConfig.OCIdleState  = TIM_OCIDLESTATE_RESET;
         ^
/home/amakarov/build/zephyr-deviance/drivers/pwm/pwm_stm32.c:131:25: error: ‘TIM_OCIDLESTATE_RESET’ undeclared (first use in this function)
  sConfig.OCIdleState  = TIM_OCIDLESTATE_RESET;
                         ^~~~~~~~~~~~~~~~~~~~~
[85/115] Building C object zephyr/kernel/CMakeFiles/kernel.dir/idle.c.obj
ninja: build stopped: subcommand failed.
```

erwango commented:
@deviance 

> pwm driver does not have support for STM32L0 timers.

Ok then, let PWM aside for now.

galak commented on boards/arm/b_l072z_lrwan1/board.h:
```diff
@@ -0,0 +1,12 @@
+#ifndef __INC_BOARD_H
+#define __INC_BOARD_H
+
+#include <soc.h>
```
remove soc.h

erwango commented:
@deviance, Please remove board.h and we'll be good. Thanks

deviance commented:
> @deviance, Please remove board.h and we'll be good. Thanks

@erwango boards.h was removed with the latest update.
